src/rgw/rgw_lc.cc

```diff
@@ -1268,27 +1291,78 @@ class LCOpAction_Transition : public LCOpAction {
     /* If bucket is versioned, create delete_marker for current version
      */
     if (oc.bucket->versioned() && oc.o.is_current() && !oc.o.is_delete_marker()) {
-      ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectExpiration);
+      ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectTransitionCurrent);
       ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") current & not delete_marker" << " versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
     } else {
-      ret = remove_expired_obj(oc.dpp, oc, true, rgw::notify::ObjectExpiration);
+      ret = remove_expired_obj(oc.dpp, oc, true, rgw::notify::ObjectTransitionNoncurrent);
       ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") not current " << "versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
     }
     return ret;
   }
 
   int transition_obj_to_cloud(lc_op_ctx& oc) {
+    int ret{0};
     /* If CurrentVersion object, remove it & create delete marker */
     bool delete_object = (!oc.tier->retain_head_object() ||
                      (oc.o.is_current() && oc.bucket->versioned()));
 
-    int ret = oc.obj->transition_to_cloud(oc.bucket, oc.tier.get(), oc.o,
-					  oc.env.worker->get_cloud_targets(), oc.cct,
-					  !delete_object, oc.dpp, null_yield);
+    /* notifications */
+    std::unique_ptr<rgw::sal::Bucket> bucket;
+    std::unique_ptr<rgw::sal::Object> obj;
+    auto& bucket_info = oc.bucket->get_info();
+    std::string version_id;
+
+    ret = oc.driver->get_bucket(nullptr, bucket_info, &bucket);
```

Comment:
what is the difference between `bucket` and `oc.bucket`?

Reply:
similar questions on `user` and `obj`?

Reply:
they are all SAL objects, not "rgw types".
for example, in line 1317 you define an `rgw::sal::Bucket*` but `oc.bucket` is also `rgw::sal::Bucket*`.
what is the difference between these objects and why do you need to allocate the local bucket in line 1322?

same question regarding `user` and `obj`.

Reply:
There were no sal types in lc_op_ctx until recently.  But additionally, creating those sal objects happens also in remove_expired_obj(), which Daniel reviewed over a year ago, so I followed the same pattern I'd used before.  It's possible this was an oversight, but it also changes more code than is being added, which I don't really like.  There is no SAL user in lc_op_ctx.

Reply:
I'll try revising both code paths, but we'll need to test more

Reply:
So, looking at remove_expired_obj(), it looks like it was a mistake by me.  I think we can just use the versions on oc, and not make copies of them.